Source versioning: Postgres, MySQL and Load generator (#647)
MaterializeInc/terraform-provider-materialize: docs/guides/materialize_source_table.md

Reviewer: This appears to be a rendered version of `templates/guides/materialize_source_table.md.tmpl` but doesn't have the header that states it's generated. We should probably add that so no-one edits this file directly.

---
page_title: "Source versioning: migrating to `materialize_source_table` Resource"
subcategory: ""
description: |-

---

# Source versioning: migrating to `materialize_source_table_{source}` Resource

In previous versions of the Materialize Terraform provider, source tables were defined within the source resource itself and were considered subsources of the source rather than separate entities.

This guide will walk you through the process of migrating your existing source table definitions to the new `materialize_source_table_{source}` resource.

> For each source type (e.g., MySQL, Postgres, etc.), you will need to create a new `materialize_source_table_{source}` resource for each table that was previously defined within the source resource. This ensures that the tables are preserved during the migration process.

Reviewer 2: ```suggestion
For each source type (e.g., MySQL, Postgres, etc.), you will need to create a new `materialize_source_table_{source}` resource for each table that was previously defined within the source resource. This ensures that the tables are preserved during the migration process. For Kafka sources, you will need to create at least one `materialize_source_table_kafka` table to hold data for the kafka topic.
```
@morsapaes might have better wording for this but I think we should be clear that this migration needs to happen for sources that previously didn't have subsources too (e.g. kafka)

  name         = "mysql_source"
  cluster_name = "cluster_name"

  mysql_connection {
    name = materialize_connection_mysql.mysql_connection.name
  }

  // Remove the table blocks from here
  - table {
  -   upstream_name        = "mysql_table1"
  -   upstream_schema_name = "shop"
  -   name                 = "mysql_table1_local"
  -
  -   ignore_columns = ["about"]
  -
  ...
}
```

This will drop the old tables from the source resources.

### Step 4: Update Terraform State

After removing the `table` blocks from your source resources, run `terraform plan` and `terraform apply` again to update the Terraform state and apply the changes.

### Step 5: Verify the Migration

After applying the changes, verify that your tables are still correctly set up in Materialize by checking the table definitions using Materialize’s SQL commands.

During the migration, you can use both the old `table` blocks and the new `materialize_source_table_{source}` resources simultaneously. This allows for a gradual transition until the old method is fully deprecated.

The same approach can be used for other source types such as Postgres, eg. `materialize_source_table_postgres`.

> ## Automated Migration Process (TBD)

Reviewer 2: nice - this is great! We will probably want to figure out how to tell them that they will be able to coordinate the 'automated' migration process with their field-engineer representative if they go down this path

---
# generated by https://github.com/hashicorp/terraform-plugin-docs
# template file: templates/guides/materialize_source_table.md.tmpl
page_title: "Source Table Migration Guide"
subcategory: ""
description: |-
  Guide for migrating to the new materialize_source_table_{source_type} resources.
---

# Source versioning: migrating to `materialize_source_table_{source_type}` Resource

In previous versions of the Materialize Terraform provider, source tables were defined within the source resource itself and were considered subsources of the source rather than separate entities.

> This guide will walk you through the process of migrating your existing source table definitions to the new `materialize_source_table_{source_type}` resource.

Reviewer 2: I'm confused on whether this guide is meant to explain how to migrate using just the manual process (which is outlined below and looks correct), or also to explain how to reconcile your terraform configuration with the results of the 'automatic' migration process that we would do for a customer with a catalog migration. Specifically, the paragraph below sounds like it is explaining what to do to handle the automatic migration process, but it doesn't explain that the manual process would then be unnecessary.

It might be simpler if this guide just explained the manual process and we had a separate one to explain the 'automatic' process. Or we could just assume any terraform user would only do the manual process. @morsapaes thoughts?

Author: Good point! I've removed some of the confusing text so that this should focus on the manual migration!

Happy to work on a follow up migration guide for the automated migration if we decide to do that!

> For each source type (e.g., MySQL, Postgres, etc.), you will need to create a new `materialize_source_table_{source_type}` resource for each table that was previously defined within the source resource. This ensures that the tables are preserved during the migration process. For Kafka sources, you will need to create at least one `materialize_source_table_kafka` table to hold data for the kafka topic.

Reviewer 2: ```suggestion
For each MySQL and Postgres source, you will need to create a new `materialize_source_table_{source_type}` resource for each table that was previously defined within the source resource. This ensures that the tables are preserved during the migration process. For Kafka sources, you will need to create a `materialize_source_table_kafka` table with the same name as the kafka source to contain the data for the kafka topic.
```

Due to limitations in the current read function, not all properties of the source tables are available when importing. To work around this, you'll need to use the `ignore_changes` lifecycle meta-argument for certain attributes that can't be read back from the state.

For example, for a Kafka source table:

```hcl
resource "materialize_source_table_kafka" "kafka_table_from_source" {
  name           = "kafka_table_from_source"
  schema_name    = "public"
  database_name  = "materialize"

  source_name = materialize_source_kafka.kafka_source.name

  include_key     = true
  include_headers = true

  envelope {
    upsert = true
  }

  lifecycle {
    ignore_changes = [
      include_key,
      include_headers,
      envelope
      ... Add other attributes here as needed
    ]
  }
}
```

This `ignore_changes` block tells Terraform to ignore changes to these attributes during subsequent applies, preventing Terraform from trying to update these values based on incomplete information from the state.

After importing, you may need to manually update these ignored attributes in your Terraform configuration to match the actual state in Materialize.

## Future Improvements

> The Kafka and Webhooks sources are currently being implemented. Once these changes, the migration process will be updated to include them.

Reviewer 2: ```suggestion
Webhooks sources have not yet been migrated to the new model. Once this changes, the migration process will be updated to include them.
```
